xrootd/xrdfuse/xrdfuse.go

```diff
@@ -0,0 +1,257 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package xrdfuse contains the implementation of the FUSE API
+// accessing a remote filesystem served over the XRootD protocol.
+package xrdfuse
+
+import (
+	"context"
+	"os"
+	"path"
+
+	"github.com/hanwen/go-fuse/fuse"
+	"github.com/hanwen/go-fuse/fuse/nodefs"
+	"github.com/hanwen/go-fuse/fuse/pathfs"
+	"github.com/pkg/errors"
+	"go-hep.org/x/hep/xrootd/client"
+	"go-hep.org/x/hep/xrootd/xrdfs"
+	"go-hep.org/x/hep/xrootd/xrdproto"
+)
+
+type FS struct {
+	pathfs.FileSystem
+	client  *client.Client
+	xrdfs   xrdfs.FileSystem
+	root    string
+	handler ErrorHandler
+}
+
+// ErrorHandler is the function which handles occurred error (e.g. logs it).
+type ErrorHandler func(error)
+
+// NewFS returns a new path.FileSystem representing the filesystem on the remote XRootD server.
+// client is a client connected to the remote XRootD server.
+// root is the path to the remote directory to be used as a root directory.
+// handler is the function which handles occurred error (e.g. logs it).
+func NewFS(client *client.Client, root string, handler ErrorHandler) *FS {
```

Comment:
perhaps we could check here whether `handler` is `nil`.
in the case it's `nil`, we could add a no-op handler, so all the `fs.handle != nil` checks that are sprinkled in `FS` and `File` could be removed.

(we should also document this in the doc-string of `NewFS`.)

alternatively, we could go with the usual optional-func design:
```go
func NewFS(client *client.Client, root string, opts ...Option) *FS { ... }

type Option func(*FS)

func WithErrorHandler(f func (error)) Option {
    return func(fs *FS) {
         fs.handler = f
    }
}
```